### ISSUE TITLE
How to build for TPU ?

### ISSUE THREAD

**Issue by janpfeifer** [openxla/xla#11599]
I'm an ML Framework ([GoMLX, for Go language](https://github.com/gomlx/gomlx)) developer using XLA (XlaBuilder for now, but at some point changing to PJRT), and I wanted to run some of my trainers in Cloud TPUs. 

How do I build XLA for TPUs ? The `configuire.py` doesn't seem to have such a `--build=TPU` option. Or is there a pre-built distribution of XLA libraries for TPUs I could use ?

Thanks!

**Comment by cheshire** [comment]
Hey,

the TPU backend is non-OSS, so you can't build it from the GH repo.

**Comment by janpfeifer** [comment]
Oh, I'm sad to hear that ... I used to see TPU related code in the repo. 

But if non-OSS, how can I use Cloud TPUs with my project ? Is there a pre-compiled binary/.so that I can link to ?

**Comment by cheshire** [comment]
Yeah there's a libtpu.so you get which gives you access: https://cloud.google.com/tpu/docs/runtimes

**Comment by janpfeifer** [comment]
> Yeah there's a libtpu.so you get which gives you access: https://cloud.google.com/tpu/docs/runtimes

Thanks. But what is inside this `libtpu.so` ? Would I have to extract the exported symbols, and reverse-engineer some `.h` files, and somehow connect the OSS XLA to it ? The page https://cloud.google.com/tpu/docs/runtimes doesn't provide any details.

**Comment by cheshire** [comment]
I think people don't access it at that level - they would use XLA (or JAX/PT) interfaces, and the implementation details would be hidden behind PjRT/StableHLO APIs.

**Comment by janpfeifer** [comment]
(Updated description of bug to clarify that my question is from the point of view of someone using XLA to create an ML framework in another language not python)

**Comment by cheshire** [comment]
Maybe the question should be then non-Python non-C++ StableHLO/PjRT/xla_builder bindings?

We've used to have quite a few other frontends: Julia, Elixir, etc, but those don't look very active these days.

**Comment by janpfeifer** [comment]
Well, I'm using xla_builder, but I'd happily change to use StableHLO/PjRT if it would enable using TPUs.

I modeled the very first version of the XLA bindings for Go on the Elixir version.

**Comment by notlober** [comment]
@janpfeifer I'm looking to build an automatic differentiation library without using front-ends like TensorFlow/JAX/PyTorch-XLA, but I'm finding information about lower-level TPU usage is practically non-existent. I see you were interested in creating an ML framework too. Do you have any insights or suggestions on how to approach this, particularly regarding TPU support? Any ideas or help would be greatly appreciated.

**Comment by janpfeifer** [comment]
hi @notlober , 

A couple of high level aspects to consider:

1. My understanding is that currently the XLA JIT compilation and optimization is behind a more well defined C-API called PJRT. It uses a `.so` (or DLL) model, the so called PJRT plugin. 

This allows for proprietary accelerators (hardware chip makers) to distribute binary plugins. 

TPU is such a proprietary plugin: you will probably find it in the Google Cloud boxes with TPU support. And presumably -- I haven't tried yet -- you can develop against a standard CPU or GPU PJRT plugin, and it will work all the same ...

I recently separated my ML framework ([GoMLX](https://github.com/gomlx/gomlx)) implementation from a Go wrapper around the PJRT (now in a [separate github project called gopjrt](https://github.com/gomlx/gopjrt)) -- I highly recommend separating these things: PJRT will give you JIT-compilation and fast execution of ML ops, and nothing else. You can then build your thing on top of it.

If you are in C++ land, you can use the PJRT wrappers already offered under github.com/openxla/xla repository -- but they are hard to compile/link. 

2. The PJRT consumes an "HLO", some intermediary proto representation of a the accelerated program you are going to execute. You still need to figure out how to translate your language/library to that.

In C++, openxla/xla offer a library called XlaBuilder: it is hard to link/compile, but it works very well. I'm wrapping that in the `gopjrt` project.

There is this "StableHLO" (and many variations) which is a textual language -- not human friendly to read/write, but enough for debugging or to generate -- which you can consider. But then you'll have to figure how to convert that to the HLO proto needed -- these tools are in openxla/xla repository somewhere, but the same concerns apply.

I hope this helps. If you can read Go (or for the Bazel's `WORKSPACE` and `BUILD` files) you may want to bootstrap your code from my repositories and change from there.

It's lots of work, but XLA runs really well, and has been very stable. I've been training many models with my framework without any issues (mostly in CPU and GPU, I haven't yet tried TPU, it's in my TODO list -- I've been postponing due to the $$$ renting the cloud boxes).

**Comment by aniruthraj** [comment]
Hi @janpfeifer ,

Thank you for raising this issue! Since there hasn't been any activity for some time, we're going to close it for now. If this issue is still relevant or you have new information to share, please feel free to reopen it. 

We appreciate your contribution and understanding!

**Comment by janpfeifer** [comment]
Thanks @aniruthraj . For the record, I've had reports of folks using my project `gopjrt` with TPUs, so all should be working fine.

Apologies I never closed the bug.

**Comment by hammer** [comment]
In case anyone else lands on this issue, to save you a few clicks: @janpfeifer has moved `gopjrt` to https://github.com/gomlx/go-xla

**Comment by janpfeifer** [comment]
Thanks @hammer ,  and let me update the state of affairs, at least for [GoMLX](https://github.com/gomlx/gomlx)/[go-xla](https://github.com/gomlx/go-xla) here:

The `go-xla` library now automatically downloads the TPU PJRT from the Jax pip package, and things just work: I develop for CPU/GPU in my consumer desktop with a graphics card, and then execute on Google Cloud on TPU (currently the cheapest ML-training/$, at least for my models) . So, not necessarily one needs to compile the TPU PJRT, just use the Jax distributed one.
